In [2]:
# Data Loading
import pandas as pd

df = pd.read_csv('loan_data_2007_2014.csv', index_col=None)

/var/folders/42/3d9r3rcx2vgd6bmrbmhv51k40000gn/T/ipykernel_57325/2366045077.py:4: DtypeWarning: Columns (20) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('loan_data_2007_2014.csv', index_col=None)


In [3]:
df.head()

Unnamed: 0       id  member_id  loan_amnt  funded_amnt  funded_amnt_inv  \
0           0  1077501    1296599       5000         5000           4975.0   
1           1  1077430    1314167       2500         2500           2500.0   
2           2  1077175    1313524       2400         2400           2400.0   
3           3  1076863    1277178      10000        10000          10000.0   
4           4  1075358    1311748       3000         3000           3000.0   

         term  int_rate  installment grade  ... total_bal_il il_util  \
0   36 months     10.65       162.87     B  ...          NaN     NaN   
1   60 months     15.27        59.83     C  ...          NaN     NaN   
2   36 months     15.96        84.33     C  ...          NaN     NaN   
3   36 months     13.49       339.31     C  ...          NaN     NaN   
4   60 months     12.69        67.79     B  ...          NaN     NaN   

  open_rv_12m open_rv_24m  max_bal_bc all_util total_rev_hi_lim inq_fi  \
0         NaN         NaN         NaN      NaN              NaN    NaN   
1         NaN         NaN         NaN      NaN              NaN    NaN   
2         NaN         NaN         NaN      NaN              NaN    NaN   
3         NaN         NaN         NaN      NaN              NaN    NaN   
4         NaN         NaN         NaN      NaN              NaN    NaN   

  total_cu_tl inq_last_12m  
0         NaN          NaN  
1         NaN          NaN  
2         NaN          NaN  
3         NaN          NaN  
4         NaN          NaN  

[5 rows x 75 columns]

# Missing Values

## Missing Values Function

In [4]:
def missing_cols(x):
        '''
        Returns a table containing in dataframe format.

                Parameters:
                        x (pandas DataFrame): DataFrame input to be inspected for the missing data
                Returns:
                        missing_value_df (pandas DataFrame): DataFrame output showing the missing data for each columns in x
        '''

        #acquiring column name, missing count and its percentage to the size of dataframe
        column_name =  x.columns
        missing_count = x.isnull().sum()
        percent_missing = missing_count * 100 / len(x)

        #creating a dataframe to store the missing row count
        missing_value_df = pd.DataFrame({
                                    'column_name': column_name,
                                    'missing_count' : missing_count,
                                    'percent_missing': percent_missing
                                    }).reset_index(drop=True)

        return missing_value_df



def missing_info(x):
        missing_value_df = missing_cols(x)
        
        missing_category = ['No Null', 'Semi Null', 'Fully Null']
        
        NoNull_count = len(missing_value_df[missing_value_df['missing_count'] == 0])
        SemiNull_count = len(missing_value_df[(
                                        (missing_value_df['missing_count'] != 0) & 
                                        (missing_value_df['percent_missing'] != 100)
                                        )])
        FullyNull_count = len(missing_value_df[missing_value_df['percent_missing'] == 100])

        CountList = [NoNull_count, SemiNull_count, FullyNull_count]
        
        NoNull_tables = missing_value_df[missing_value_df['missing_count'] == 0]['column_name'].tolist()
        SemiNull_tables = missing_value_df[(
                                        (missing_value_df['missing_count'] != 0) & 
                                        (missing_value_df['percent_missing'] != 100)
                                        )]['column_name'].tolist()
        FullyNull_tables = missing_value_df[missing_value_df['percent_missing'] == 100]['column_name'].tolist()

        Tables = [NoNull_tables, SemiNull_tables, FullyNull_tables]

        missing_info = pd.DataFrame({
                                    'missing_category': missing_category,
                                    'table_number' : CountList,
                                    'table_list': Tables
                                    }).reset_index(drop=True)

        return missing_value_df, missing_info




## Missing Data Understanding

In [5]:
missings_cal, missing_stat = missing_info(df)
missings_cal

column_name  missing_count  percent_missing
0         Unnamed: 0              0         0.000000
1                 id              0         0.000000
2          member_id              0         0.000000
3          loan_amnt              0         0.000000
4        funded_amnt              0         0.000000
..               ...            ...              ...
70          all_util         466285       100.000000
71  total_rev_hi_lim          70276        15.071469
72            inq_fi         466285       100.000000
73       total_cu_tl         466285       100.000000
74      inq_last_12m         466285       100.000000

[75 rows x 3 columns]

In [6]:
missing_stat

missing_category  table_number  \
0          No Null            35   
1        Semi Null            23   
2       Fully Null            17   

                                          table_list  
0  [Unnamed: 0, id, member_id, loan_amnt, funded_...  
1  [emp_title, emp_length, annual_inc, desc, titl...  
2  [annual_inc_joint, dti_joint, verification_sta...

In [7]:
#showing only missing columns
missings_cal[missings_cal['missing_count'] != 0]

column_name  missing_count  percent_missing
11                    emp_title          27588         5.916553
12                   emp_length          21008         4.505399
14                   annual_inc              4         0.000858
20                         desc         340302        72.981546
22                        title             20         0.004289
26                  delinq_2yrs             29         0.006219
27             earliest_cr_line             29         0.006219
28               inq_last_6mths             29         0.006219
29       mths_since_last_delinq         250351        53.690554
30       mths_since_last_record         403647        86.566585
31                     open_acc             29         0.006219
32                      pub_rec             29         0.006219
34                   revol_util            340         0.072917
35                    total_acc             29         0.006219
46                 last_pymnt_d            376         0.080637
48                 next_pymnt_d         227214        48.728567
49           last_credit_pull_d             42         0.009007
50   collections_12_mths_ex_med            145         0.031097
51  mths_since_last_major_derog         367311        78.773926
54             annual_inc_joint         466285       100.000000
55                    dti_joint         466285       100.000000
56    verification_status_joint         466285       100.000000
57               acc_now_delinq             29         0.006219
58                 tot_coll_amt          70276        15.071469
59                  tot_cur_bal          70276        15.071469
60                  open_acc_6m         466285       100.000000
61                   open_il_6m         466285       100.000000
62                  open_il_12m         466285       100.000000
63                  open_il_24m         466285       100.000000
64           mths_since_rcnt_il         466285       100.000000
65                 total_bal_il         466285       100.000000
66                      il_util         466285       100.000000
67                  open_rv_12m         466285       100.000000
68                  open_rv_24m         466285       100.000000
69                   max_bal_bc         466285       100.000000
70                     all_util         466285       100.000000
71             total_rev_hi_lim          70276        15.071469
72                       inq_fi         466285       100.000000
73                  total_cu_tl         466285       100.000000
74                 inq_last_12m         466285       100.000000

In [8]:
#showing only missing columns and not entirely missing (100%)
missings_cal[((missings_cal['missing_count'] != 0) & (missings_cal['percent_missing'] != 100))].sort_values(by='percent_missing', ascending=False)

column_name  missing_count  percent_missing
30       mths_since_last_record         403647        86.566585
51  mths_since_last_major_derog         367311        78.773926
20                         desc         340302        72.981546
29       mths_since_last_delinq         250351        53.690554
48                 next_pymnt_d         227214        48.728567
71             total_rev_hi_lim          70276        15.071469
59                  tot_cur_bal          70276        15.071469
58                 tot_coll_amt          70276        15.071469
11                    emp_title          27588         5.916553
12                   emp_length          21008         4.505399
46                 last_pymnt_d            376         0.080637
34                   revol_util            340         0.072917
50   collections_12_mths_ex_med            145         0.031097
49           last_credit_pull_d             42         0.009007
28               inq_last_6mths             29         0.006219
31                     open_acc             29         0.006219
35                    total_acc             29         0.006219
27             earliest_cr_line             29         0.006219
26                  delinq_2yrs             29         0.006219
57               acc_now_delinq             29         0.006219
32                      pub_rec             29         0.006219
22                        title             20         0.004289
14                   annual_inc              4         0.000858

### TOP 5 missing columns

| Table Name | Table Description | Missing percentage |
| --- | --- | --- |
| mths_since_last_record | The number of months since the last public record. | 84.6% |
| mths_since_last_major_derog | Months since most recent 90-day or worse rating. | 78.8% |
| desc | Loan description provided by the borrower. | 73% |
| mths_since_last_delinq | The number of months since the borrower's last delinquency. | 53.7% |
| next_pymnt_d | Next scheduled payment date | 48.7% |


<b> Hypothesis based on description </b>


- <b>mths_since_last_record</b> \
No activity/input of public record for the user 


- <b>mths_since_last_major_derog</b>\
No input/record for derogatory of the user


- <b>desc</b>\
User give no description for their loans


- <b>mths_since_last_delinq</b>\
No input/record for delinquency of the user


- <b>next_pymnt_d</b>\
No more payments needed (Check 'loan_status' column)